Develop

AndreiBrehuescu opened the pull request:
Done half of chapter 8.
Problems: When I press the delete product button, it redirects me to the main page but I need to refresh it to see the results.

brehua committed:
Created the project

brehua committed:
Update index.tsx

brehua committed:
Chapter 8 working

brehua committed:
Formatted using prittier

brehua committed:
update

brehua committed:
Updates on chapter 8

brehua committed:
delete debug logs

AndreiSold commented:
The first thing that popped into my eyes was the commit messages list:

- Created the project    (great)
- Update index.tsx       (what did you update?)
- Chapter 8 working    (what is chapter 8 about if I am unaware of the training you did?)
- Formatted using prittier   (this one is nice, maybe something to tell us that is was a project wide format (**Formatted entire project using prettier**)
- update   (what did you update?)
- Updates on chapter 8   (what did you update and what is chapter 8?)
- delete debug logs    (nice, the only improvement here would be to decide to write the commit messages either starting with a caps or lower case so they would be uniform)

AndreiSold commented on frontend/online-shop/src/App.css:
```diff
@@ -0,0 +1,38 @@
+  text-align: center;
+}
+
+.App-logo {
```
Do you still use those classes? If not you can clean this `.css` folder

AndreiBrehuescu replied:
You are right, fixed

AndreiSold commented on frontend/online-shop/src/App.tsx:
```diff
@@ -0,0 +1,26 @@
+export const App = () => {
+  return (
+    <ThemeProvider theme={theme}>
+      <div>
```
Is this div helping you with anything?

AndreiBrehuescu replied:
It was useless, fixed

AndreiSold commented on frontend/online-shop/src/App.tsx:
```diff
@@ -0,0 +1,26 @@
+    <ThemeProvider theme={theme}>
+      <div>
+        <Header />
+        <main>
```
Same with the main

AndreiBrehuescu replied:
It was useless, fixed

AndreiSold commented on frontend/online-shop/src/App.tsx:
```diff
@@ -0,0 +1,26 @@
+      <div>
+        <Header />
+        <main>
+          <Switch>
```
What if the user is directly accessing a path that you don't have in your switch? For example http://localhost:4200/asdfg. For such cases you need to have a wildcard route that catches all the paths that do not fit with any other declared route and return a 404 page.

AndreiBrehuescu replied:
Added a path for unknown  requests

AndreiSold commented on frontend/online-shop/src/App.tsx:
```diff
@@ -0,0 +1,26 @@
+            <Route exact path="/">
+              <Redirect to="/products" />
+            </Route>
+            <Route path="/product/:number" component={ProductDetails} />
```
You should be more specific with the path parameters. Either /product/:id or /product/:productId.

AndreiBrehuescu replied:
Now its more specific

AndreiSold commented on frontend/online-shop/src/Header.tsx:
```diff
@@ -0,0 +1,29 @@
+import { useStyles } from "./styles/tableStyles";
+
+const Header = () => (
+  <header>
```
You don't neccessarily need to use vanilla html tags. You can achieve the results even by only using Material UI components such as Grid, Box, etc

AndreiBrehuescu replied:
Refactored to use only Material UI

AndreiSold commented on frontend/online-shop/src/Header.tsx:
```diff
@@ -0,0 +1,29 @@
+      <Grid container spacing={3}>
+        <Grid item xs={12} sm={6}>
+          <Link to={{ pathname: `/` }}>
+            <IconButton className={useStyles().buttonStyle}>
```
You should extract the `useStyles()` in a const above the return because the way you do it you call the useStyles() method on each jsx element. If you extract it you will make the call only once and then retrieve the classes from there.

AndreiBrehuescu replied:
Fixed

AndreiSold commented on frontend/online-shop/src/Header.tsx:
```diff
@@ -0,0 +1,29 @@
+            </IconButton>
+          </Link>
+        </Grid>
+        <Grid item xs={12} sm={6}>
```
Excellent use of the grid and its breakpoints

AndreiSold commented on frontend/online-shop/src/components/ProductDetails.tsx:
```diff
@@ -0,0 +1,87 @@
+
+const ProductDetails = (props: any) => {
+  const product = props.location.state.product;
+  const classes = useStyles();
```
Here you used useStyles the right way

AndreiSold commented on frontend/online-shop/src/components/ProductDetails.tsx:
```diff
@@ -0,0 +1,87 @@
+import ProductService from "../services/ProductService";
+
+const ProductDetails = (props: any) => {
+  const product = props.location.state.product;
```
As this is a component that is directly tied to a route you shouldn't use props with it. Indeed when you come to this page from the products list page you will be able to send the product and retrieve it from the props. How about when you try to access directly http://localhost:4200/products/1? Also there is a method called useParams() that can retrieve the path parameters object without using props

AndreiSold replied:
You should retrieve the product id from the path and make a service call to retrieve the product with the respective id.

AndreiSold commented on frontend/online-shop/src/components/ProductDetails.tsx:
```diff
@@ -0,0 +1,87 @@
+                    <IconButton
+                      className={classes.buttonStyle}
+                      onClick={() => {
+                        productService.deleteProductById(product.id);
```
Nice

AndreiSold commented on frontend/online-shop/src/components/ProductDetails.tsx:
```diff
@@ -0,0 +1,87 @@
+};
+
+export default ProductDetails;
+function componentDidMount() {
```
This shouldn't be here anymore as you are now working with functional components.

AndreiBrehuescu replied:
Fixed

AndreiSold commented on frontend/online-shop/src/components/Products.tsx:
```diff
@@ -0,0 +1,60 @@
+import ProductService from "../services/ProductService";
+import React from "react";
+
+export interface ProductEntityProps {
```
Excellent use of an interface for the props

AndreiSold commented on frontend/online-shop/src/components/Products.tsx:
```diff
@@ -0,0 +1,60 @@
+}
+
+export const TableProducts = (props: ProductEntityProps) => {
+  var productService = new ProductService();
```
Try to avoid using var, use const instead

AndreiSold replied:
Here's why: https://janakachathuranga.medium.com/why-we-use-const-and-avoid-var-3528ee550fb4

AndreiBrehuescu replied:
Understood , refactored

AndreiSold commented on frontend/online-shop/src/components/Products.tsx:
```diff
@@ -0,0 +1,60 @@
+  const [data, setData] = React.useState<any[]>([]);
+
+  React.useEffect(() => {
+    result.then((res) => {
```
On line 21 you are retrieving a promise and storing it within a variable. This is not usually recommended. You should make the call directly in the `useEffect` hook. Also you should have a `.catch` callback in case the server returns an error and you must handle it. The alternative to `.then` and `.catch` callbacks would be to use `await`.

AndreiBrehuescu replied:
Understood, refactored to use redux instead

AndreiSold commented on frontend/online-shop/src/components/Products.tsx:
```diff
@@ -0,0 +1,60 @@
+
+  const classes = useStyles();
+  const items = data.map((product) => (
+    <tr className={classes.trStyle}>
```
React needs a key on the item elements within a map result. You could wrap everything with a `<Box key={product.id}>`

AndreiBrehuescu replied:
Added the key value to the <tr> tag because using the box compressed everything in a cell only.

AndreiSold commented on frontend/online-shop/src/components/Products.tsx:
```diff
@@ -0,0 +1,60 @@
+      <td className={classes.thtdStyle}>{product.name}</td>
+      <td className={classes.thtdStyle}>{product.price}</td>
+      <td key={product.id}>
+        <Link to={{ pathname: `/product/${product.id}`, state: { product } }}>
```
As I mentioned within the productDetails component you should avoid sending the product as a prop and retrieve the product id from the path so you can make a service call to retrieve it.

AndreiSold commented on frontend/online-shop/src/services/ProductService.ts:
```diff
@@ -0,0 +1,21 @@
+  constructor() {}
+
+  getProducts() {
+    const result = axios.get("http://localhost:8080/products", {});
```
The second parameter is optional, you can just place the path as the first parameter

AndreiBrehuescu replied:
Fixed

AndreiSold commented on frontend/online-shop/src/styles/tableStyles.tsx:
```diff
@@ -0,0 +1,49 @@
+import { createStyles, Theme } from "@material-ui/core";
+
+export const useStyles = makeStyles((theme: Theme) =>
+  createStyles({
```
Great use of the theme

AndreiSold commented on frontend/online-shop/src/styles/tableTheme.tsx:
```diff
@@ -0,0 +1,29 @@
+      main: "#009879",
+    },
+    secondary: {
+      //tbody tr background color
```
Remove such comments. Leave only the comments that are documenting complicate functions or business logic

AndreiBrehuescu replied:
Fixed

AndreiSold reviewed (changes requested):
Great job! The progress is really nice. I left you some comments regarding the commit messages you have written, some suggestions, good practices and refactoring ideas. Keep up the great work!

brehua committed:
Started Chapter9

Created store
Create interfaces for the product, category and supplier
Created and updated the service for product
Created reducers for the main page product table
Created the action for get all product
Tried to connect all of these using redux.

brehua committed:
Pull reuqest chamges

Deleted unused css file
Added 404Not found route for unrecognized requests

brehua committed:
Update

Fixed pull request changed,
Refactored code to use redux and saga
Modified ProductInterface so match the backend JSON type
Created recuder, saga and store files to manage the redux
Created update page where you can  modify the price and the name of an existing product

brehua committed:
Cart functionalities update

Created cart actions file, cart saga and cart reducer
Added a new state to the store
Combined the rudecers and sagas for the store
Everything still on the testing phase

brehua committed:
Update

Added interfaces to for OrderModel to send the request to the backend
Updated reducers and sagas for Cart state in the sore
Created the cart and checkout page
Added address field for the order
Created confirmation dialog for succes or fail in order creation
Updated cart actions and removed all project side debugging logs.
